### PR TITLE
tidb: add document for external timestamp read

### PR DESCRIPTION
**Pull request by YangKeao**
Signed-off-by: YangKeao <yangkeao@chunibyo.icu>

### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### What is changed, added or deleted? (Required)

Add a document for external timestamp read and related system variables.

### Which TiDB version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [x] v6.4 (TiDB 6.4 versions)
- [ ] v6.3 (TiDB 6.3 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from: https://github.com/pingcap/docs/pull/10783
- Other reference link(s): https://github.com/pingcap/tidb/issues/38274

### CODE REVIEW

**Comment by easonn7** [comment]
LGTM

**Review by lcwangchao — APPROVED** [review]

File: system-variables.md [hunk 1]
```diff
@@ -2705,3 +2705,17 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
 - 是否持久化到集群：是
 - 默认值：`ON`
 - 这个变量用于控制计算窗口函数时是否采用高精度模式。
+
+### tidb_external_ts
```
Comment (Oreoxmt) [inline system-variables.md:2709]:
```suggestion
### tidb_external_ts <span class="version-mark">从 v6.4.0 版本开始引入</span>
```

File: system-variables.md [hunk 2]
```diff
@@ -2705,3 +2705,17 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
+
+- 作用域: GLOBAL
+- 是否持久化到集群：是
+- 默认值: `0`
+- 当 `tidb_enable_external_ts_read` 设置为 `ON` 时, TiDB 会依据该时间戳读取历史数据。
+
+### tidb_enable_external_ts_read
```
Comment (Oreoxmt) [inline system-variables.md:2716]:
```suggestion
### tidb_enable_external_ts_read <span class="version-mark">从 v6.4.0 版本开始引入</span>
```

File: system-variables.md [hunk 3]
```diff
@@ -2705,3 +2705,17 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
 - 是否持久化到集群：是
 - 默认值：`ON`
 - 这个变量用于控制计算窗口函数时是否采用高精度模式。
+
+### tidb_external_ts
+
+- 作用域: GLOBAL
+- 是否持久化到集群：是
+- 默认值: `0`
+- 当 `tidb_enable_external_ts_read` 设置为 `ON` 时, TiDB 会依据该时间戳读取历史数据。
```
Comment (Oreoxmt) [inline system-variables.md:2714]:
```suggestion
- 作用域：GLOBAL
- 是否持久化到集群：是
- 默认值：`0`
- 当 [`tidb_enable_external_ts_read`](#tidb_enable_external_ts_read-从-v640-版本开始引入) 设置为 `ON` 时，TiDB 会依据该时间戳读取历史数据。
```

File: system-variables.md [hunk 4]
```diff
@@ -2705,3 +2705,17 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
+
+### tidb_enable_external_ts_read
+
+- 作用域: SESSION | GLOBAL
+- 是否持久化到集群：是
+- 默认值: `OFF`
+- 当此变量设置为 `ON` 时, TiDB 会读取 `tidb_external_ts` 指定时间戳前的历史数据。
```
Comment (Oreoxmt) [inline system-variables.md:2721]:
```suggestion
- 作用域：SESSION | GLOBAL
- 是否持久化到集群：是
- 默认值：`OFF`
- 当此变量设置为 `ON` 时，TiDB 会读取 [`tidb_external_ts`](#tidb_external_ts-从-v640-版本开始引入) 指定时间戳前的历史数据。
```

File: tidb-external-ts.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+summary: 了解如何通过系统变量 `tidb_external_ts` 读取历史数据。
+---
+
+# 通过系统变量 `tidb_external_ts` 读取历史数据
+
+为了支持出读取历史版本数据，TiDB 从 6.4 版本起引入了一个新的系统变量 `tidb_external_ts`。本文档介绍如何通过该系统变量读取历史数据，其中包括具体的操作流程。
+
+> **警告：**
+>
+> 目前 Stale Read 特性无法和 TiFlash 一起使用。如果你在查询时将 `tidb_enable_external_ts_read` 设置为 `ON` 并且 TiDB 可能从 TiFlash 副本读取数据，你可能会遇到 `ERROR 1105 (HY000): stale requests require tikv backend` 报错信息。
```
Comment (Oreoxmt) [inline tidb-external-ts.md:12]:
```suggestion
> 目前 Stale Read 特性无法和 TiFlash 一起使用。如果你在查询时将 [`tidb_enable_external_ts_read`](/system-variables.md#tidb_enable_external_ts_read-从-v640-版本开始引入) 设置为 `ON` 并且 TiDB 可能从 TiFlash 副本读取数据，你可能会遇到 `ERROR 1105 (HY000): stale requests require tikv backend` 报错信息。
```

File: tidb-external-ts.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+>
+> 要解决该问题，你需要为使用 Stale Read 特性的查询禁用 TiFlash 副本。要禁用 TiFlash 副本，你可以使用以下任一方法：
+>
+> + 通过设置变量来禁用 TiFlash 副本 `set session tidb_isolation_read_engines='tidb,tikv'`。
```
Comment (Oreoxmt) [inline tidb-external-ts.md:16]:
```suggestion
> + 通过设置变量来禁用 TiFlash 副本 `SET SESSION tidb_isolation_read_engines='tidb,tikv'`。
```

File: tidb-external-ts.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+> + 使用 [hint](/optimizer-hints.md#read_from_storagetiflasht1_name--tl_name--tikvt2_name--tl_name-) 强制 TiDB 从 TiKV 读取数据。
```
Comment (Oreoxmt) [inline tidb-external-ts.md:17]:
```suggestion
> + 使用 [`READ_FROM_STORAGE`](/optimizer-hints.md#read_from_storagetiflasht1_name--tl_name--tikvt2_name--tl_name-) 强制 TiDB 从 TiKV 读取数据。
```

File: tidb-external-ts.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+## 场景介绍
+
+通过配置让 TiDB 能够读取某一固定时间点的历史数据对于 TiCDC 等数据同步工具非常有用。在数据同步工具完成了某一时间点前的数据同步之后，可以通过设置下游 TiDB 的 `tidb_external_ts` 系统变量，使得下游 TiDB 的请求能够读取到该时间点前的数据。这将避免在同步过程中，下游 TiDB 读取到尚未完全同步而不一致的数据。
+
+## 示例
+
+系统变量 `tidb_external_ts` 用于指定该功能启用时，读取历史数据使用的时间戳。
```
Comment (Oreoxmt) [inline tidb-external-ts.md:25]:
```suggestion
系统变量 [`tidb_external_ts`](/system-variables.md#tidb_external_ts-从-v640-版本开始引入) 用于指定该功能启用时，读取历史数据使用的时间戳。
```

File: tidb-external-ts.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+系统变量 `tidb_enable_external_ts_read` 控制着是否在当前会话或全局启用该功能。默认值为 `OFF`，这意味着该功能关闭，并且 `tidb_external_ts` 没有作用。当该变量被全局地设置为 `ON`，所有的请求都将读取到 `tidb_external_ts` 指定时间之前的历史数据。如果 `tidb_enable_external_ts_read` 仅在某一会话被设置为 `ON`，则只有该会话中的请求会读取到历史数据。
+
+当 `tidb_enable_external_ts_read` 被设置为 `ON` 时，TiDB 会进入只读模式。任何写请求都会失败并且返回错误 `ERROR 1836 (HY000): Running in read-only mode`.
+
+以下是一个使用该功能的示例：
+
+```sql
+create table t (c int);
```
Comment (Oreoxmt) [inline tidb-external-ts.md:34]:
```suggestion
CREATE TABLE t (c INT);
```

File: tidb-external-ts.md [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+```
+
+```
+Query OK, 0 rows affected (0.01 sec)
+```
+
+```sql
+insert into t values (1), (2), (3);
```
Comment (Oreoxmt) [inline tidb-external-ts.md:42]:
```suggestion
INSERT INTO t VALUES (1), (2), (3);
```

File: tidb-external-ts.md [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+```
+
+```
+Query OK, 3 rows affected (0.00 sec)
+```
+
+查看表中的数据:
+
+```sql
+select * from t;
```
Comment (Oreoxmt) [inline tidb-external-ts.md:52]:
```suggestion
SELECT * FROM t;
```

File: tidb-external-ts.md [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+|    2 |
+|    3 |
++------+
+3 rows in set (0.00 sec)
+```
+
+将 `tidb_external_ts` 设置为 `@@tidb_current_ts`:
+
+```sql
+start transaction;set global tidb_external_ts=@@tidb_current_ts;commit;
```
Comment (Oreoxmt) [inline tidb-external-ts.md:69]:
```suggestion
START TRANSACTION;
SET GLOBAL tidb_external_ts = @@tidb_current_ts;
COMMIT;
```

File: tidb-external-ts.md [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+```
+
+插入新的一行:
+
+```sql
+insert into t values (4);
```
Comment (Oreoxmt) [inline tidb-external-ts.md:75]:
```suggestion
INSERT INTO t VALUES (4);
```

File: tidb-external-ts.md [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+```
+
+```
+Query OK, 1 row affected (0.001 sec)
+```
+
+确认新的一行已经被插入:
+
+```sql
+select * from t;
```
Comment (Oreoxmt) [inline tidb-external-ts.md:85]:
```suggestion
SELECT * FROM t;
```

File: tidb-external-ts.md [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+|    4 |
++------+
+4 rows in set (0.00 sec)
+```
+
+然而，因为 `tidb_external_ts` 被设置为插入这一行之前的时间，在将 `tidb_enable_external_ts_read` 启动后，将读取不到新插入的行:
+
+```sql
+set tidb_enable_external_ts_read=ON;
+select * from t;
```
Comment (Oreoxmt) [inline tidb-external-ts.md:104]:
```suggestion
SET tidb_enable_external_ts_read = ON;
SELECT * FROM t;
```

File: stale-read.md [hunk 16]
```diff
@@ -22,3 +22,5 @@ TiDB 提供语句级别与会话级别的 Stale Read 使用方式，具体使用
     - 指定时间范围：如需 TiDB 读取在一个时间范围内尽可能新的数据并且不破坏隔离级别，你可以指定一个时间范围。在指定时间范围内，TiDB 会选择一个合适的时间戳，该时间戳能保证所访问的副本上不存在开始于这个时间戳之前且还没有提交的相关事务，即能保证在所访问的可用副本上可执行读取操作而且不会被阻塞。要使用该方式，请参阅 [`AS OF TIMESTAMP` 语法](/as-of-timestamp.md#语法方式)文档和该文档中 [`TIDB_BOUNDED_STALENESS` 函数](/as-of-timestamp.md#语法方式)部分的介绍。
 - 会话级别：
     - 指定时间范围：在会话级别中，如需 TiDB 在后续的查询中读取一个时间范围内尽可能新的数据并且不破坏隔离级别，你可以通过设置一个 session 变量 `tidb_read_staleness` 来指定一个时间范围。要使用该方式，请参阅[通过系统变量 `tidb_read_staleness` 读取历史数据](/tidb-read-staleness.md)。
+
+除此以外，你也可以通过设置系统变量 `tidb_external_ts` 来在某一会话或全局范围读取某一时间点前的历史数据。要使用该方式，请参阅[通过系统变量 `tidb_external_ts` 读取历史数据](/tidb-external-ts.md)。
```
Comment (Oreoxmt) [inline stale-read.md:26]:
```suggestion
    - 指定时间点：如需在全局或全局范围读取某一时间点前的历史数据，你可以通过设置一个系统变量 `tidb_external_ts` 来指定一个时间点。要使用该方式，请参阅[通过系统变量 `tidb_external_ts` 读取历史数据](/tidb-external-ts.md)。
```

File: tidb-external-ts.md [hunk 17]
```diff
@@ -0,0 +1,118 @@
+---
+title: 通过系统变量 `tidb_external_ts` 读取历史数据
+summary: 了解如何通过系统变量 `tidb_external_ts` 读取历史数据。
+---
+
+# 通过系统变量 `tidb_external_ts` 读取历史数据
+
+为了支持出读取历史版本数据，TiDB 从 6.4 版本起引入了一个新的系统变量 `tidb_external_ts`。本文档介绍如何通过该系统变量读取历史数据，其中包括具体的操作流程。
```
Comment (Oreoxmt) [inline tidb-external-ts.md:8]:
```suggestion
为了支持读取历史版本数据，TiDB 从 v6.4.0 起引入了一个新的系统变量 [`tidb_external_ts`](/system-variables.md#tidb_external_ts-从-v640-版本开始引入)。本文档介绍如何通过该系统变量读取历史数据，其中包括具体的操作流程。
```

File: tidb-external-ts.md [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+
+## 场景介绍
+
+通过配置让 TiDB 能够读取某一固定时间点的历史数据对于 TiCDC 等数据同步工具非常有用。在数据同步工具完成了某一时间点前的数据同步之后，可以通过设置下游 TiDB 的 `tidb_external_ts` 系统变量，使得下游 TiDB 的请求能够读取到该时间点前的数据。这将避免在同步过程中，下游 TiDB 读取到尚未完全同步而不一致的数据。
+
+## 示例
+
+系统变量 [`tidb_external_ts`](/system-variables.md#tidb_external_ts-从-v640-版本开始引入) 用于指定该功能启用时，读取历史数据使用的时间戳。
+
+系统变量 `tidb_enable_external_ts_read` 控制着是否在当前会话或全局启用该功能。默认值为 `OFF`，这意味着该功能关闭，并且 `tidb_external_ts` 没有作用。当该变量被全局地设置为 `ON`，所有的请求都将读取到 `tidb_external_ts` 指定时间之前的历史数据。如果 `tidb_enable_external_ts_read` 仅在某一会话被设置为 `ON`，则只有该会话中的请求会读取到历史数据。
```
Comment (Oreoxmt) [inline tidb-external-ts.md:27]:
```suggestion
系统变量 [`tidb_enable_external_ts_read`](/system-variables.md#tidb_enable_external_ts_read-从-v640-版本开始引入) 控制着是否在当前会话或全局启用该功能。默认值为 `OFF`，这意味着该功能关闭，并且 `tidb_external_ts` 没有作用。当该变量被全局地设置为 `ON`，所有的请求都将读取到 `tidb_external_ts` 指定时间之前的历史数据。如果 `tidb_enable_external_ts_read` 仅在某一会话被设置为 `ON`，则只有该会话中的请求会读取到历史数据。
```

File: stale-read.md (hunk 16 above)
Comment (YangKeao) [reply stale-read.md:26]: 一个问题是它不只是会话级别，而是既可以会话级别也可以全局。所以放在 “会话级别” 下有点不太合适

File: tidb-external-ts.md (hunk 18 above)
Comment (YangKeao) [reply tidb-external-ts.md]: Fixed

File: tidb-external-ts.md (hunk 17 above)
Comment (YangKeao) [reply tidb-external-ts.md]: Fixed.

**Review by Oreoxmt — APPROVED** [review]
LGTM

**Review by shichun-0415 — APPROVED** [review]
Rest LGTM

File: system-variables.md [hunk 19]
```diff
@@ -966,6 +966,20 @@ MPP 是 TiFlash 引擎提供的分布式计算框架，允许节点之间的数
...
+
+- 作用域：SESSION | GLOBAL
+- 是否持久化到集群：是
+- 默认值：`OFF`
+- 当此变量设置为 `ON` 时，TiDB 会读取 [`tidb_external_ts`](#tidb_external_ts-从-v640-版本开始引入) 指定时间戳前的历史数据。
+
+### `tidb_external_ts` <span class="version-mark">从 v6.4.0 版本开始引入</span>
+
+- 作用域：GLOBAL
+- 是否持久化到集群：是
```
[the remark this answers is not in the document]
Comment (YangKeao) [reply system-variables.md:980]: Fixed.

**Comment by Oreoxmt** [comment]
DNM until https://github.com/pingcap/tidb/pull/38276 is merged

**Pull request merged**